### ISSUE TITLE
As a user, I want to be able to post data for the WRES to use in evaluations that is not necessary tied directly to a specific evaluation

### ISSUE THREAD

**Issue by epag** [NOAA-OWP/wres#164]
---

Author Name: **Hank** (Hank)
Original Redmine Issue: 127251, https://vlab.noaa.gov/redmine/issues/127251
Original Date: 2024-03-01

---

Feel free to change the title. The idea is for a user to be able to post data they want to evaluate, get back what would be needed to refer to that data in an evaluation, and then perform evaluations using that data.  This would be a departure from the current, data-direct approach that allows for an evaluation to be posted and then data supporting that evaluation to be posted before the evaluation proceeds.  

This ticket can be resolved once a solution is identified (likely using a third party library of some sort), appropriate IT resources are obtained, and the solution is implemented and put in place. We'll definitely need this in staging before production, but its unclear if deploying it to -dev is necessary. Perhaps a reduced version of it? That can be discussed.

Setting this to a high priority and leaving in the back log.

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **Hank** (Hank)
Original Date: 2024-03-01T19:50:13Z

---

Related to #125894, because it would help to address the missing value issue reported by Anna.

Related to #127239, because, if we are going to be making significant IT requests in staging and production, it may make sense to bundle them together and handle them all at one. Blue-green will be a big ask of IT. Having a data/object/whatever store supporting WRES would also be a big ask at least in terms of disk space.

Hank

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **Hank** (Hank)
Original Date: 2024-03-01T19:58:46Z

---

Hank:

Research this next week:

https://hub.docker.com/r/bitnami/minio

Hank

**Comment by HankHerr-NOAA** [comment]
I'm going to do just a bit of preliminary research related to this ticket, specifically looking into minio as an option since that was recommended previously. I'm starting with a simple video,

https://www.youtube.com/watch?v=kckykvbaZog&t=161s

that shows how to run an instance in Docker.  

This ticket will cover any code changes required to support the use of minio or whatever tool is selected. Once that tool is selected, we'll need a VLab ticket for changes to the COWRES environment to support use of that tool. But, for now, I'm just investigating options.

Hank

**Comment by HankHerr-NOAA** [comment]
Just doing some basic tests in -dev03, and I can't get past this error:

```[hank.herr@nwcal-wres-dev03 minio_test]$ docker run -p 9000:9000 -p 9090:9090 --name minio -v /mnt/wres_share/minio_test/data:/data -e "MINIO_ROOT_USER=minioadmon" -e "MINIO_ROOT_PASSWORD=minioadmin" quay.io/minio/minio server /data --console_address ":9090"
Incorrect Usage: flag provided but not defined: -console_address
```

Web searches for that error have not bee fruitful. The only reference I can find appears to be for Kubernetes and minio operator. That reference talks about incompatible versions, but I'm pretty sure it does not apply since operator appears to only apply to Kubernetes, which I'm not using (obviously).  

If I start the container without the `--console_address` argument, the container comes up and stays up, and, hopping into the container, I can confirm that '--console_address' is a valid argument for `minio server`. 

Still investigating,

Hank

**Comment by james-d-brown** [comment]
There's a weird mix of short and long option names (`-` and `--`) in the above command. Perhaps use all of one or the other, just to make it more readable. The error message itself seems to imply that the long form (`--`) was seen as a shortcut ('`'), but I have no idea...

**Comment by HankHerr-NOAA** [comment]
Yes, that's what the message implies.  That particular option is only available in long form. All of the other options are Docker options, not minio, so they are unrelated.  I appear to be able to get minion running without a console if I leave out the argument, but I can only confirm that by the errors Isee when I attempt to use `curl` to talk to it.

Hank

**Comment by james-d-brown** [comment]
Oh, you are using an underscore! It isn't an underscore, it's a dash.

**Comment by james-d-brown** [comment]
Ha ha, how frustrating :)

**Comment by HankHerr-NOAA** [comment]
Why do I do that to myself? :) Thanks for spotting it!

Before I just saw your message, I was able to use curl to list the buckets, of which there were none, successfully.  Yippee! I did it based on what I saw in this stack overflow:

https://stackoverflow.com/questions/69773176/minio-create-and-list-a-bucket-with-curl

Now that I have a console, let me see if I can interact it,

Hank

**Comment by HankHerr-NOAA** [comment]
I was able to upload files through the console and see them placed on the machine. After cycling the containers with my test data directory mounted to the worker container, I was able to get WRES to "see" the files, but it can't read them directly because minio uses S3. We would need to resurrect our old S3 bucket reader.  Still, this is all promising.

Obviously, there are a lot of questions that still need to be answered:

1. Is Minio the best option? What are the alternatives? Which do we prefer?

2. Do we want to allow for uses to directly upload files, and let WRES act as a middle man, checking the validity and readability of files before posting them on the file system?

3. How do we establishing expiration dates and an algorithm to remove files when the disk is near full?

4. How do we setup TLS?  

5. And the low level stuff, like how best to configure the minio instance, deploy it, repo control it, etc. 

We'll need to discuss this further. I'm now in a meeting with Jim and won't return to this until later this afternoon or tomorrow. Thanks,

Hank

**Comment by HankHerr-NOAA** [comment]
Recording some notes from yesterday on starting and interacting with minio, and some takeaways...

Here is the command to start the test minio instance through which I was able to create a bucket and upload data:

docker run -p 9000:9000 -p 9090:9090 --name minio -v /mnt/wres_share/minio_test/data:/data -e "MINIO_ROOT_USER=omitted" -e "MINIO_ROOT_PASSWORD=omitted" quay.io/minio/minio server /data --console-address ":9090"

The bucket was created through the console via browser (port 9090, as shown above), and I was able to go into the bucket and upload data. The data then appeared in the file system under `/mnt/wres_share/minio_test/data`, as shown above. However, files cannot just be read as is; S3 or whatever must be used. I did not test this, but perhaps a `curl` request could be prepared to pull the data file as an example. 

Also worth noting: network issues prevented an upload of three of the files. I'm not sure if that's going to be a thing in production, but it was in my test.  Best approach would probably be to package all of the files into a .tgz and upload that, but I chose to upload the files that support a smoke test, and 3 did not transfer successfully.

A `curl` to obtain the listing of a bucket looks like this:

curl -k -H "Host: $URL" -H "Date: $DATE" -H "Content-Type: $CONTENT_TYPE" -H "Authorization: AWS $USERNAME:$SIGNATURE" http://$URL:9000/wrestest

The variables were setup as follows:

```
USERNAME=omitted       # Matched what I used to setup minio
PASSWORED=omitted      # Matched what I used during setup
MINIO_PATH="/"
DATE=$(date -R --utc)
CONTENT_TYPE='application/zstd'
SIG_STRING="GET\n\n${CONTENT_TYPE}\n${DATE}\n${MINIO_PATH}"
SIGNATURE=`echo -en ${SIG_STRING} | openssl sha1 -hmac ${PASSWORD} -binary | base64`
```

The takeaway: Getting a list of what is in the bucket requires a username and password. That imlies,

- we need to handle communications with the minio instance on behalf of our users (because we know the username and password to use),
- arbitrary users will need to get their own credentials, OR
- we need to disable to the need for username and password. 

I think we would need to create an endpoint to support user interactions, specifically allowing for files to be posted, but, again, we aren't even sure minio is the solution. See my previous comment with some questions that need answering.

More in the dev call,

Hank

**Comment by HankHerr-NOAA** [comment]
Answer the first question about alternatives, for the last few minutes of work today, I read about CEPH. After reading through the Google AI response, I found this comparison:

https://stackshare.io/stackups/ceph-vs-minio

From that, it appear MinIO may be better for small, simple solutions, which I think is what we would want. CEPH might be more geared for larger scale solutions.

That was just from an overview of a few minutes. I'll take a deeper look tomorrow.

Hank
